kroxylicious/src/test/java/io/kroxylicious/proxy/internal/filter/FetchResponseTransformationFilterTest.java

```diff
@@ -0,0 +1,219 @@
+
+        when(context.createByteBufferOutputStream(bufferInitialCapacity.capture())).thenAnswer(
+                (Answer<ByteBufferOutputStream>) invocation -> {
+                    Object[] args = invocation.getArguments();
+                    Integer size = (Integer) args[0];
+                    return new ByteBufferOutputStream(size);
+                });
+    }
+
+    @Test
+    void filterHandlesResponseBasedOnTopicNames() throws Exception {
+
+        var fetchResponse = new FetchResponseData();
+        fetchResponse.responses().add(createFetchableTopicResponseWithOneRecord(RECORD_KEY, ORIGINAL_RECORD_VALUE).setTopic(TOPIC_NAME)); // Version 12
+
+        var stage = filter.onFetchResponse(fetchResponse.apiKey(), new ResponseHeaderData(), fetchResponse, context);
+        assertThat(stage).isCompleted();
+
+        var filteredResponse = (FetchResponseData) stage.toCompletableFuture().get().message();
+        var filteredRecords = responseToRecordStream(filteredResponse).toList();
+        assertThat(filteredRecords)
+                .withFailMessage("unexpected number of records in the filter response")
+                .hasSize(1);
+
+        var filteredRecord = filteredRecords.get(0);
+        assertThat(decodeUtf8Value(filteredRecord))
+                .withFailMessage("expected record value to have been transformed")
+                .isEqualTo(EXPECTED_TRANSFORMED_RECORD_VALUE);
+    }
+
+    @Test
+    void filterHandlesResponseBasedOnTopicIds() throws Exception {
+
+        var fetchResponse = new FetchResponseData();
+        // Version 13 switched to topic id rather than topic names.
+        fetchResponse.responses().add(createFetchableTopicResponseWithOneRecord(RECORD_KEY, ORIGINAL_RECORD_VALUE).setTopicId(TOPIC_ID));
+
+        var metadataResponse = new MetadataResponseData();
+        metadataResponse.topics().add(new MetadataResponseData.MetadataResponseTopic().setTopicId(TOPIC_ID).setName(TOPIC_NAME));
+
+        when(context.sendRequest(anyShort(), isA(MetadataRequestData.class)))
+                .thenReturn(CompletableFuture.completedStage(metadataResponse));
+
+        var stage = filter.onFetchResponse(fetchResponse.apiKey(), new ResponseHeaderData(), fetchResponse, context);
+        assertThat(stage).isCompleted();
+
+        var filteredResponse = (FetchResponseData) stage.toCompletableFuture().get().message();
+
+        // verify that the response now has the topic name
+        assertThat(filteredResponse.responses())
+                .withFailMessage("expected same number of topics in the response")
+                .hasSameSizeAs(fetchResponse.responses())
+                .withFailMessage("expected topic response to have been augmented with topic name")
+                .anyMatch(ftr -> Objects.equals(ftr.topic(), TOPIC_NAME))
+                .withFailMessage("expected topic response to still have the topic id")
+                .anyMatch(ftr -> Objects.equals(ftr.topicId(), TOPIC_ID));
```

Comment:
nit: Given we have 3 different messages should this be 3 different assertions?

Reply:
Sonar seems to complain if I don't coalesce assertions that have the same subject.

Reply:
*sigh* let's leave it I'm not inclined to fight sonar